pkg/syncer/upsync/upsync_controller.go

```diff
@@ -0,0 +1,437 @@
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), controllerName),
+		upstreamClient:   upstreamClient,
+		downstreamClient: downstreamClient,
+		getDownstreamLister: func(gvr schema.GroupVersionResource) (cache.GenericLister, error) {
+			informers, notSynced := ddsifForDownstream.Informers()
+			informer, ok := informers[gvr]
+			if !ok {
+				if shared.ContainsGVR(notSynced, gvr) {
+					return nil, fmt.Errorf("informer for gvr %v not synced in the downstream informer factory", gvr)
+				}
+				return nil, fmt.Errorf("gvr %v should be known in the downstream informer factory", gvr)
+			}
+			return informer.Lister(), nil
+		},
+		listDownstreamNamespacesByLocator: func(jsonLocator string) ([]*unstructured.Unstructured, error) {
+			nsInformer, err := ddsifForDownstream.ForResource(namespaceGVR)
+			if err != nil {
+				return nil, err
+			}
+			return indexers.ByIndex[*unstructured.Unstructured](nsInformer.Informer().GetIndexer(), syncerindexers.ByNamespaceLocatorIndexName, jsonLocator)
+		},
+		getUpstreamUpsyncerLister: func(gvr schema.GroupVersionResource) (kcpcache.GenericClusterLister, error) {
+			informers, notSynced := ddsifForUpstreamUpyncer.Informers()
+			informer, ok := informers[gvr]
+			if !ok {
+				if shared.ContainsGVR(notSynced, gvr) {
+					return nil, fmt.Errorf("informer for gvr %v not synced in the upstream upsyncer informer factory", gvr)
+				}
+				return nil, fmt.Errorf("gvr %v should be known in the downstream upstream upsyncer informer factory", gvr)
+			}
+			return informer.Lister(), nil
+		},
+		getUpsyncedGVRs: func() ([]schema.GroupVersionResource, error) {
+			informers, notSynced := ddsifForUpstreamUpyncer.Informers()
+			var result []schema.GroupVersionResource
+			for k := range informers {
+				result = append(result, k)
+			}
+			if len(notSynced) > 0 {
+				return result, fmt.Errorf("informers not synced in the upstream upsyncer informer factory for gvrs %v", notSynced)
+			}
+			return result, nil
+		},
+
+		syncTargetName:        syncTargetName,
+		syncTargetClusterName: syncTargetClusterName,
+		syncTargetUID:         syncTargetUID,
+		syncTargetKey:         syncTargetKey,
+	}
+	logger := logging.WithReconciler(syncerLogger, controllerName)
+
+	emptyMap := map[string]interface{}{}
+	ddsifForDownstream.AddEventHandler(ddsif.GVREventHandlerFuncs{
+		AddFunc: func(gvr schema.GroupVersionResource, obj interface{}) {
+			if gvr == namespaceGVR {
+				return
+			}
+			new, ok := obj.(*unstructured.Unstructured)
+			if !ok {
+				utilruntime.HandleError(fmt.Errorf("resource should be a *unstructured.Unstructured, but was %T", obj))
+				return
+			}
+			if new.GetLabels()[workloadv1alpha1.ClusterResourceStateLabelPrefix+syncTargetKey] != string(workloadv1alpha1.ResourceStateUpsync) {
+				return
+			}
+			status := new.UnstructuredContent()["status"]
+			c.enqueueDownstream(gvr, new, logger, status != nil && !equality.Semantic.DeepEqual(status, emptyMap))
```

Comment:
technically I think an emptyMap could be dirty as well, imagine we have to sync the status because in upstream it is not empty.

Reply:
I agree on a technical POV, but OTOH we don't expect and manage the case where the status (or more generally the content of the upstream resource) would be changed or removed upstream, since upsynced resources are expected to be completely owned by downstream (this will be enforced through an admission controller at the KCP level at some point).
So I'm not sure it really makes sense to take this case into account.

Reply:
it's unnecessarily brittle now, even has extra code here for that. Why not simple and more reliable? What is the downside? Also note that this is not about modifcations upstream. IMO the code now is just wrong. A downstream objects can get an empty status. You must sync that.

Reply:
I rolled back to syncing empty statuses.